rpcs3/rpcs3qt/main_window.cpp

```diff
@@ -3531,20 +3539,99 @@ void main_window::SetIconSizeActions(int idx) const
 		ui->setIconSizeLargeAct->setChecked(true);
 }
 
-void main_window::RemoveDiskCache()
+void main_window::RemoveHDD1Caches()
+{
+	if (fs::remove_all(rpcs3::utils::get_hdd1_dir() + "caches", false))
+	{
+		QMessageBox::information(this, tr("HDD1 Caches Removed"), tr("HDD1 caches successfully removed"));
+	}
+	else
+	{
+		QMessageBox::warning(this, tr("Error"), tr("Could not remove HDD1 caches"));
+	}
+}
+
+void main_window::RemoveAllCaches()
+{
+	if (QMessageBox::question(this, tr("Confirm Removal"), tr("Remove all caches?")) != QMessageBox::Yes)
+		return;
+
+	const std::string cache_base_dir = rpcs3::utils::get_cache_dir();
+	u64 caches_count = 0;
+	u64 caches_removed = 0;
+
+	for (const game_info& game : m_game_list_frame->GetGameInfo()) // Loop on discovered games
+	{
+		if (game && qstr(game->info.category) != cat::cat_ps3_os && fs::exists(cache_base_dir + game->info.serial)) // If not OS category and cache exists
+		{
+			caches_count++;
+
+			if (fs::remove_all(cache_base_dir + game->info.serial))
+			{
+				caches_removed++;
+			}
+		}
+	}
+
+	if (caches_count == caches_removed)
+	{
+		QMessageBox::information(this, tr("Caches Removed"), tr("%0 cache(s) successfully removed").arg(caches_removed));
+	}
+	else
+	{
+		QMessageBox::warning(this, tr("Error"), tr("Could not remove %0 of %1 cache(s)").arg(caches_count - caches_removed).arg(caches_count));
+	}
+
+	RemoveHDD1Caches();
+}
+
+void main_window::RemoveSavestates()
 {
-	const std::string cache_dir = rpcs3::utils::get_hdd1_dir() + "/caches";
+	if (QMessageBox::question(this, tr("Confirm Removal"), tr("Remove savestates?")) != QMessageBox::Yes)
+		return;
 
-	if (fs::remove_all(cache_dir, false))
+	if (fs::remove_all(fs::get_config_dir() + "savestates", false))
 	{
-		QMessageBox::information(this, tr("Cache Cleared"), tr("Disk cache was cleared successfully"));
+		QMessageBox::information(this, tr("Savestates Removed"), tr("Savestates successfully removed"));
 	}
 	else
 	{
-		QMessageBox::warning(this, tr("Error"), tr("Could not remove disk cache"));
+		QMessageBox::warning(this, tr("Error"), tr("Could not remove savestates"));
 	}
 }
 
+void main_window::CleanupGamesList()
+{
+	if (QMessageBox::question(this, tr("Confirm Removal"), tr("Remove no more discoverable games from games list \"games.yml\" file?")) != QMessageBox::Yes)
+		return;
+
+	// List of serials (title id) to remove in "games.yml" file (if any)
+	std::vector<std::string> serials_to_remove_from_yml{};
+
+	for (const auto& [serial, path] : Emu.GetGamesConfig().get_games()) // Loop on games list "games.yml" file
+	{
+		bool found = false;
+
+		for (const game_info& game : m_game_list_frame->GetGameInfo()) // Loop on discovered games
```

Comment:
Why? Just check if the path in the games.yml exists?

Reply:
it's more complete this way no? if there us some corruption that prevents it from being added to the game list it would be detected here.
Also, this loop is cheaper than filesystem call at nearly all times.

Reply:
I compare the entries (title id) in games.yml with the games properly discovered (and displayed in the games list frame). The path could be present but the content could be corrupted so it will be also a zombie